src/mds/MDCache.cc

```diff
@@ -13585,7 +13585,8 @@ void MDCache::upkeep_main(void)
         if (active_with_clients) {
           trim_client_leases();
         }
-        if (is_open()) {
+	/* Wait rejoin ack to finish */
+        if (is_open() && rejoin_ack_gather.empty()) {
```

Comment:
Perhaps we should defer trimming whenever the cluster is degraded?

Reply:
Circling back: my concern here is the endless special cases we seem to introduce in code like this. Preventing cache trim when a rank is unavailable is probably not viable but I wonder how we can clean this up. Perhaps if there is a rank in `up:rejoin` then do not trim cache, with an explanation.

Reply:
Yeah, sounds reasonable to me. Let me go through the code again to improve it. 
Thanks @batrick

Reply:
> Perhaps we should defer trimming whenever the cluster is degraded?

maybe we start using `mds->is_cluster_degraded()` everywhere i guess?

Reply:
Sorry I didn't walk back my bad suggestion strongly enough: I don't think we can stop trimming the cache just because the cluster is degraded. It is not uncommon to have completely disjoint subtrees on the ranks (think CSI PVs) which means clients will not necessarily have a session with the not-active rank. That means those clients will happily continue interacting with the active ranks and filling up their caches. The assumption otherwise would be that all clients would eventually have some request block on the not-active rank which would bring all file system I/O to a halt (thereby preventing the active ranks caches from overflowing) until the cluster is no longer degraded.

So, I think a targeted fix is necessary but I'm apprehensive about even testing if there is a rank in `up:rejoin`. Is there a sensible way to pin the subtree if we're waiting on an ack?

Reply:
> Sorry I didn't walk back my bad suggestion strongly enough: I don't think we can stop trimming the cache just because the cluster is degraded. It is not uncommon to have completely disjoint subtrees on the ranks (think CSI PVs) which means clients will not necessarily have a session with the not-active rank. That means those clients will happily continue interacting with the active ranks and filling up their caches. The assumption otherwise would be that all clients would eventually have some request block on the not-active rank which would bring all file system I/O to a halt (thereby preventing the active ranks caches from overflowing) until the cluster is no longer degraded.
> 
> So, I think a targeted fix is necessary but I'm apprehensive about even testing if there is a rank in `up:rejoin`. Is there a sensible way to pin the subtree if we're waiting on an ack?

This seems a better approach. Let me check and confirm it. Thanks @batrick

Reply:
I am afraid we couldn't just pin the subtree, the PINing means it will only pin it in the MDCache, couldn't pin it into the `subtrees[]` and doing this will make the code more complicated to maintain.

We can just check the `rejoin_done` instead.

Reply:
> Sorry I didn't walk back my bad suggestion strongly enough: I don't think we can stop trimming the cache just because the cluster is degraded. 

Fair enough.

It is not uncommon to have completely disjoint subtrees on the ranks (think CSI PVs) which means clients will not necessarily have a session with the not-active rank.

What's a `not-active` rank? I understand `active` and other MDS states, but maybe `not-active` is a terminology I'm not familiar with or does it mean all active ranks other than the one a client has a session to?

Reply:
> What's a not-active rank?

I mean it's a rank recovering from failure (either no assigned MDS or state < up:active)